File: java/samples/batch-mode/src/main/java/com/myscript/iink/samples/batchmode/MainActivity.java
/*
 * Copyright (c) MyScript. All rights reserved.
 */

package com.myscript.iink.samples.batchmode;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.DisplayMetrics;
import android.util.Log;

import com.google.gson.Gson;
import com.myscript.iink.Configuration;
import com.myscript.iink.ContentPackage;
import com.myscript.iink.ContentPart;
import com.myscript.iink.Editor;
import com.myscript.iink.Engine;
import com.myscript.iink.MimeType;
import com.myscript.iink.PointerEvent;
import com.myscript.iink.PointerEventType;
import com.myscript.iink.Renderer;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    // To enable text recognition for a specific language,
    // see also: https://developer.myscript.com/support/recognition-assets
    private static final String language = "en_US";
    private static final String packageName = "package.iink";
    private static final String exportFileName = "export";
    // Choose type of content ("Text", "Math", "Diagram", "Raw Content")
    private static String partType = "Text";
    private Engine engine;
    private Editor editor;
    private DisplayMetrics displayMetrics;
    private ContentPackage contentPackage;
    private ContentPart contentPart;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        engine = IInkApplication.getEngine();

        // Configure recognition
        Configuration conf = engine.getConfiguration();
        String confDir = "zip://" + getPackageCodePath() + "!/assets/conf";

        conf.setStringArray("configuration-manager.search-path", new String[]{confDir});
        conf.setString("content-package.temp-folder", getFilesDir().getPath() + File.separator + "tmp");
        conf.setString("lang", language);

        // Configure the engine to disable guides (recommended)
        conf.setBoolean("text.guides.enable", false);

        // Create a renderer with a null render target
        displayMetrics = getResources().getDisplayMetrics();
        Renderer renderer = engine.createRenderer(displayMetrics.xdpi, displayMetrics.ydpi, null);

        // Create the editor
        editor = engine.createEditor(renderer);

        // The editor requires a font metrics provider and a view size *before* calling setPart()
        editor.setFontMetricsProvider(null);
        editor.setViewSize(displayMetrics.widthPixels, displayMetrics.heightPixels);

        // Load the pointerEvents from the right .json file depending of the part type
        PointerEvent[] pointerEvents = loadPointerEvents();

        try {
            // Create a new package
            contentPackage = engine.createPackage(packageName);

            // Create a new part
            contentPart = contentPackage.createPart(partType);
        } catch (IOException e) {
            Log.e(TAG, "Failed to open package \"" + packageName + "\"", e);
        } catch (IllegalArgumentException e) {
            Log.e(TAG, "Failed to open package \"" + packageName + "\"", e);
        }

        // Associate editor with the new part
        editor.setPart(contentPart);

        // Feed the editor
        editor.pointerEvents(pointerEvents, false);

        // Export the result of the recognition into a file
        export();

        finish();
    }

    @Override
    protected void onDestroy() {
        // IInkApplication has the ownership, do not close here
        engine = null;

        super.onDestroy();
    }

    void export() {
        MimeType mimeType = MimeType.TEXT;

        switch (partType) {
            case "Math":
                mimeType = MimeType.LATEX;
                break;

            case "Diagram":
                mimeType = MimeType.SVG;
                break;

            case "Raw Content":
                mimeType = MimeType.JIIX;
                break;

            default:
                break;
        }

        // Exported file is stored in the Virtual SD Card : "Android/data/com.myscript.iink.batchmode/files"
        File file = new File(getExternalFilesDir(null), File.separator + exportFileName + mimeType.getFileExtensions());
        editor.waitForIdle();

        try {
            editor.export_(null, file.getAbsolutePath(), mimeType, null);
        } catch (Exception e) {
            e.printStackTrace();
        }

        editor.setPart(null);
        contentPart.close();
        contentPackage.close();

        try {
            engine.deletePackage(packageName);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private PointerEvent[] loadPointerEvents() {
        PointerEvent[] pointerEvents = null;
        String pointerEventsPath = "conf/pointerEvents/" + partType.toLowerCase() + "/pointerEvents.json";

        if (partType.equals("Text")) {
            pointerEventsPath = "conf/pointerEvents/" + partType.toLowerCase() + "/" + language + "/pointerEvents.json";
        }

        try {
            // Loading the content of the pointerEvents JSON file
            InputStream inputStream = getResources().getAssets().open(pointerEventsPath);

            // Mapping the content into a JsonResult class
            JsonResult jsonResult = new Gson().fromJson(new InputStreamReader(inputStream), JsonResult.class);

            int pointerEventsCount = 0;

            // Calculating the number of pointerEvents
            for (Stroke stroke : jsonResult.getStrokes()) {
                pointerEventsCount += stroke.getX().length;
            }

            // Allocating the size of the array of pointerEvents
            pointerEvents = new PointerEvent[pointerEventsCount];

            int i = 0;

            for (Stroke stroke : jsonResult.getStrokes()) {
                float[] strokeX = stroke.getX();
                float[] strokeY = stroke.getY();
                long[] strokeT = stroke.getT();
                float[] strokeP = stroke.getP();
                int length = stroke.getX().length;

                for (int j = 0; j < length; j++) {
                    PointerEvent pointerEvent = new PointerEvent();
                    pointerEvent.pointerType = stroke.getPointerType();
                    pointerEvent.pointerId = stroke.getPointerId();

                    if (j == 0) {
                        pointerEvent.eventType = PointerEventType.DOWN;
                    } else if (j == length - 1) {
                        pointerEvent.eventType = PointerEventType.UP;
                    } else {
                        pointerEvent.eventType = PointerEventType.MOVE;
                    }

                    // Transform the x and y coordinates of the stroke from mm to px
                    // This is needed to be adaptive for each device
                    pointerEvent.x = (strokeX[j] / 25.4f) * displayMetrics.xdpi;
                    pointerEvent.y = (strokeY[j] / 25.4f) * displayMetrics.ydpi;

                    pointerEvent.t = strokeT[j];
                    pointerEvent.f = strokeP[j];
                    pointerEvents[i++] = pointerEvent;
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return pointerEvents;
    }
}

File: java/samples/batch-mode-kt/src/main/java/com/myscript/iink/samples/batchmode/MainActivity.kt
package com.myscript.iink.samples.batchmode

import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.RadioGroup
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import androidx.core.content.res.ResourcesCompat
import com.google.gson.Gson
import com.myscript.iink.ConversionState
import com.myscript.iink.Editor
import com.myscript.iink.MimeType
import com.myscript.iink.PointerEvent
import com.myscript.iink.PointerEventType
import com.myscript.iink.PointerType
import com.myscript.iink.Renderer
import com.myscript.iink.uireferenceimplementation.FontMetricsProvider
import com.myscript.iink.uireferenceimplementation.FontUtils
import com.myscript.iink.uireferenceimplementation.ImageLoader
import com.myscript.iink.uireferenceimplementation.ImagePainter
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileInputStream
import java.io.InputStreamReader
import java.nio.file.Files

class MainActivity : AppCompatActivity() {

    private lateinit var renderer : Renderer
    private lateinit var editor: Editor
    private lateinit var imagePainter: ImagePainter

    private var useCustomInputFile = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        val engine = IInkApplication.getEngine()

        if (engine == null) {
            // The certificate provided is incorrect, you need to use the one provided by MyScript
            AlertDialog.Builder(this)
                .setTitle( getString(R.string.app_error_invalid_certificate_title))
                .setMessage( getString(R.string.app_error_invalid_certificate_message))
                .setPositiveButton(com.myscript.iink.app.common.R.string.dialog_ok, null)
                .show()
            finishAndRemoveTask() // be sure to end the application
            return
        }

        // Configure recognition
        engine.apply {
            configuration.apply {
                val confDir = "zip://${application.packageCodePath}!/assets/conf"
                setStringArray("configuration-manager.search-path", arrayOf(confDir))
                setString("content-package.temp-folder", application.cacheDir.absolutePath)
                // To enable text recognition for a specific language,
                setString("lang", language);
                // Configure the engine to disable guides (recommended in batch mode)
                setBoolean("text.guides.enable", false);
            }
        }

        // At creation we have to pre initialise the editor with dpi and screen size

        // Create a renderer with a null render target
        val displayMetrics = resources.displayMetrics
        renderer = engine.createRenderer(displayMetrics.xdpi, displayMetrics.ydpi, null).apply {
            setViewOffset(0.0f, 0.0f)
            viewScale = 1.0f
        }

        // Create the editor
        editor = engine.createEditor(renderer, engine.createToolController()).apply {
            // The editor requires a font metrics provider and a view size *before* calling setPart()
            val typefaceMap = mutableMapOf<String, Typeface>()
            setFontMetricsProvider(FontMetricsProvider(displayMetrics, typefaceMap))
            setViewSize(displayMetrics.widthPixels, displayMetrics.heightPixels)
        }

        // Create a image painter to render in png
        imagePainter = ImagePainter().apply {
            setImageLoader(ImageLoader(editor))
            // load fonts
            val typefaceMap = provideTypefaces()
            setTypefaceMap(typefaceMap)
            editor.setFontMetricsProvider(FontMetricsProvider(applicationContext.resources.displayMetrics, typefaceMap))
            editor.theme = (".math {font-family: STIX;}")
        }

        useCustomInputFile = false

        findViewById<AppCompatButton>(R.id.batch_sample_pick_file).setOnClickListener { _ -> pickFile() }

        findViewById<AppCompatButton>(R.id.batch_sample_remove_file).setOnClickListener(this::removeFile)

        findViewById<AppCompatButton>(R.id.batch_sample_execute).setOnClickListener(this::execute)
    }

    private fun provideTypefaces(): Map<String, Typeface> {
        val typefaces = FontUtils.loadFontsFromAssets(application.assets) ?: mutableMapOf()
        // Map key must be aligned with the font-family used in theme.css
        val myscriptInterFont = ResourcesCompat.getFont(application, R.font.myscriptinter)
        if (myscriptInterFont != null) {
            typefaces["MyScriptInter"] = myscriptInterFont
        }
        val stixFont = ResourcesCompat.getFont(application, R.font.stix)
        if (stixFont != null) {
            typefaces["STIX"] = stixFont
        }
        return typefaces
    }

    private fun pickFile() {
        pickFileLauncher.launch(Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            type = "application/*"
            addCategory(Intent.CATEGORY_OPENABLE)
        })
    }

    private fun removeFile(removeButton : View) {
        useCustomInputFile = false
        removeButton.visibility = View.GONE
        findViewById<AppCompatButton>(R.id.batch_sample_pick_file).visibility = View.VISIBLE
        findViewById<TextView>(R.id.batch_sample_file_description).visibility = View.VISIBLE
    }

    private val pickFileLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) {
            it.data?.data?.also { uri ->
                val contentResolver = applicationContext.contentResolver
                contentResolver.openInputStream(uri)?.use { inputStream ->
                    val workingFile = File(applicationContext.cacheDir, customInputFileName)
                    workingFile.delete()
                    Files.copy(inputStream, workingFile.toPath())
                }

                useCustomInputFile = true
                findViewById<AppCompatButton>(R.id.batch_sample_remove_file).visibility = View.VISIBLE
                findViewById<AppCompatButton>(R.id.batch_sample_pick_file).visibility = View.GONE
                findViewById<TextView>(R.id.batch_sample_file_description).visibility = View.GONE
            }
        }

    @OptIn(DelicateCoroutinesApi::class)
    private fun execute(executeButton: View) {
        val partTypeGroup = findViewById<RadioGroup>(R.id.batch_sample_part_type_group)
        val partType = when(partTypeGroup.checkedRadioButtonId) {
            R.id.batch_sample_part_type_math -> "Math"
            R.id.batch_sample_part_type_diagram -> "Diagram"
            R.id.batch_sample_part_type_raw -> "Raw Content"
            else -> "Text"
        }

        val modeGroup = findViewById<RadioGroup>(R.id.batch_sample_mode_group)
        val incremental = when(modeGroup.checkedRadioButtonId) {
            R.id.batch_sample_mode_incremental -> true
            else -> false
        }

        val outputGroup = findViewById<RadioGroup>(R.id.batch_sample_output_group)
        val pngOutput = when(outputGroup.checkedRadioButtonId) {
            R.id.batch_sample_output_png -> true
            else -> false
        }

        val outputStyleGroup = findViewById<RadioGroup>(R.id.batch_sample_output_style_group)
        val convert = when(outputStyleGroup.checkedRadioButtonId) {
            R.id.batch_sample_output_style_converted -> true
            else -> false
        }

        val progress = findViewById<ProgressBar>(R.id.batch_sample_progress)

        progress.visibility = View.VISIBLE
        executeButton.isEnabled = false
        partTypeGroup.isEnabled = false
        modeGroup.isEnabled = false
        outputGroup.isEnabled = false
        outputStyleGroup.isEnabled = false

        GlobalScope.launch(Dispatchers.Main) {
            val outputFile = withContext(Dispatchers.IO) {
                process(partType, incremental, pngOutput, convert)
            }

            progress.visibility = View.GONE
            executeButton.isEnabled = true
            partTypeGroup.isEnabled = true
            modeGroup.isEnabled = true
            outputGroup.isEnabled = true
            outputStyleGroup.isEnabled = true

            if (!outputFile.exists()) {
                Toast.makeText(applicationContext, "An error occurred when exporting", Toast.LENGTH_LONG).show()
                return@launch
            }

            exportedFilePath = outputFile.absolutePath

            saveFileLauncher.launch(Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
                type = "application/*"
                addCategory(Intent.CATEGORY_OPENABLE)
                putExtra(Intent.EXTRA_TITLE, outputFile.name)
            })
        }
    }

    private var exportedFilePath: String? = null
    private val saveFileLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) {
            exportedFilePath?.let { exportedFilePath ->
                val exportedFile = File(exportedFilePath)
                if (exportedFile.exists()) {
                    it.data?.data?.also { uri ->
                        val contentResolver = applicationContext.contentResolver
                        contentResolver.openOutputStream(uri)?.use { outputStream ->
                            Files.copy(exportedFile.toPath(), outputStream)
                            this.exportedFilePath = null
                        }
                    }
                }
            }
        }

    override fun onDestroy() {
        super.onDestroy()

        editor.close()
        renderer.close()
    }

    private fun process(partType: String, incremental: Boolean, renderToPNG: Boolean, convert: Boolean): File {
        val engine = requireNotNull(IInkApplication.getEngine())

        // Create a new package
        val contentPackage =  engine.createPackage(iinkPackageName)
        // Create a new part
        val contentPart = contentPackage.createPart(partType)
        // Associate editor with the new part
        editor.part = contentPart

        // Now we can process pointer events and feed the editor with an array of Pointer Events loaded from the json file
        loadAndFeedPointerEvents(partType, incremental)

        // Choose the right mimeType to export according to the partType chosen
        var mimeType = MimeType.PNG
        if(!renderToPNG) {
            when (partType) {
                typeOfPart[0] -> mimeType = MimeType.TEXT  // Text
                typeOfPart[1] -> mimeType = MimeType.LATEX // Math
                typeOfPart[2] -> mimeType = MimeType.SVG   // Diagram
                typeOfPart[3] -> mimeType = MimeType.JIIX  // Raw Content
                else -> {}
            }
        }

        editor.waitForIdle()

        if (convert) {
            editor.convert(editor.rootBlock, ConversionState.DIGITAL_EDIT)
        }
        val outputFile = File(applicationContext.filesDir, "$exportFileName${mimeType.fileExtensions}")
        editor.export_(editor.rootBlock, outputFile.absolutePath, mimeType, if (mimeType.isImage) imagePainter else null)

        // Closing after using
        editor.part = null
        contentPart.close()
        contentPackage.close()
        engine.deletePackage(packageName)

        return outputFile
    }

    private fun loadAndFeedPointerEvents(partType: String, incremental: Boolean = false) {
        // Loading the content of the pointerEvents JSON file
        val customInputFile = File(applicationContext.cacheDir, customInputFileName)
        val inputStream =  if (useCustomInputFile && customInputFile.exists()) {
            FileInputStream(customInputFile)
        } else {
            val partTypeLowercase = partType.lowercase()
            val pointerEventsPath = if (partTypeLowercase == "text") {
                "conf/pointerEvents/$partTypeLowercase/$language/pointerEvents.json"
            } else {
                "conf/pointerEvents/$partTypeLowercase/pointerEvents.json"
            }

            resources.assets.open(pointerEventsPath)
        }

        // Mapping the content into a JsonResult class
        val jsonResult = Gson().fromJson(InputStreamReader(inputStream), JsonResult::class.java)

        // Add each element to a list
        val pointerEventsList = mutableListOf<PointerEvent>()

        val xdpi = resources.displayMetrics.xdpi
        val ydpi = resources.displayMetrics.ydpi

        jsonResult.getStrokes()?.forEach { stroke ->
            val strokeX = stroke.x
            val strokeY = stroke.y
            val strokeT = stroke.t
            val strokeP = stroke.p
            val length = stroke.x.size

            for (index in 0 until length) {
                // Transform the x and y coordinates of the stroke from mm to px
                // This is needed to be adaptive for each device
                val x = strokeX[index] / 25.4f * xdpi
                val y = strokeY[index] / 25.4f * ydpi

                if (incremental) {
                    // In incremental mode we send data direct to the editor
                    when (index) {
                        0 -> editor.pointerDown(x, y, strokeT[index], strokeP[index], PointerType.PEN, 1)
                        length - 1 -> editor.pointerUp(x, y, strokeT[index], strokeP[index], PointerType.PEN, 1)
                        else -> editor.pointerMove(x, y, strokeT[index], strokeP[index], PointerType.PEN, stroke.pointerId)
                    }
                } else {
                    // In batch mode we keep data in a array
                    pointerEventsList += PointerEvent().apply {
                        pointerType = stroke.pointerType
                        pointerId = stroke.pointerId
                        eventType = when (index) {
                            0 -> PointerEventType.DOWN
                            length - 1 -> PointerEventType.UP
                            else -> PointerEventType.MOVE
                        }
                        this.x = x
                        this.y = y
                        t = strokeT[index]
                        f = strokeP[index]
                    }
                }
            }
        }
        if (!incremental) {
            editor.pointerEvents(pointerEventsList.toTypedArray(), false)
        }
    }

    companion object {
        private const val TAG = "MainActivity"

        private const val customInputFileName = "customInputFile"

        // /!\ Warning use the real MyScript name of part as this string will be used for part creation
        private val typeOfPart = listOf("Text", "Math", "Diagram", "Raw Content")
        private const val iinkPackageName = "package.iink"
        private const val exportFileName = "export"
        private const val language = "en_US"
    }
}